Prepare for packer release 1.15.1 (#1999)
hashicorp/web-unified-docs: content/packer/v1.15.x/content/docs/provisioners/hcp-sbom.mdx

---
description: |
  The `hcp-sbom` Packer provisioner uploads a CycloneDX- or SPDX JSON-formatted software bill of materials record to HCP Packer. Learn how to use the `hcp-sbom` provisioner.
page_title: hcp-sbom provisioner reference
# START AUTO GENERATED METADATA, DO NOT EDIT
created_at: 2025-11-17T10:53:47-06:00
last_modified: 2025-11-17T10:53:47-06:00
# END AUTO GENERATED METADATA
---

⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️
> [!IMPORTANT]  
> **Documentation Update:** Product documentation previously located in `/website` has moved to the [`hashicorp/web-unified-docs`](https://github.com/hashicorp/web-unified-docs) repository, where all product documentation is now centralized. Please make contributions directly to `web-unified-docs`, since changes to `/website` in this repository will not appear on developer.hashicorp.com.
⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️

<BadgesHeader>
  <PluginBadge type="official"/>
</BadgesHeader>

# `hcp-sbom` provisioner

> The `hcp-sbom` provisioner uploads software bill of materials (SBOM) files from artifacts built by Packer to HCP Packer. You must format SBOM files you want to upload as JSON and follow either the [SPDX](https://spdx.github.io/spdx-spec/latest) or [CycloneDX](https://cyclonedx.org/) specification. HCP Packer ties these SBOM files to the version of the artifact that Packer builds.

Reviewer: ```suggestion
The `hcp-sbom` provisioner uploads software bill of materials (SBOM) files from artifacts built by Packer to HCP Packer and associates them with individual artifact versions. You must format SBOM files you want to upload as JSON and follow either the [SPDX](https://spdx.github.io/spdx-spec/latest) or [CycloneDX](https://cyclonedx.org/) specification.
```

2. **Automatic SBOM Generation**: Automatically generate an SBOM by downloading and running an SBOM generation tool on the remote host. By default, Packer uses [Syft](https://github.com/anchore/syft) for SBOM generation, but you can specify a custom tool if needed.

## Basic Example

The following example uploads an SBOM from the local `/tmp` directory and stores a copy at `./sbom/sbom_cyclonedx.json` on the local machine.

<Tabs>
<Tab heading="HCL2">

```hcl
provisioner "hcp-sbom" {
  source      = "/tmp/sbom_cyclonedx.json"
  destination = "./sbom/sbom_cyclonedx.json"
  sbom_name   = "sbom-cyclonedx"
}
```

</Tab>
<Tab heading="JSON">

```json
{
  "type": "hcp-sbom",
  "source": "/tmp/sbom_cyclonedx.json",
  "destination": "./sbom/sbom_cyclonedx.json",
  "sbom_name": "sbom-cyclonedx"
}
```

</Tab>
</Tabs>

## Automatic SBOM Generation

The provisioner can automatically generate SBOMs by downloading and running an SBOM generation tool on the remote host. By default, Packer downloads and uses [Syft](https://github.com/anchore/syft) for SBOM generation. This eliminates the need for manual tool installation.

> ### Basic Auto-Generation Example

Reviewer: ```suggestion
The following example configures the provisioner to download Syft and automatically generate an SBOM, associate it with the HCP Packer artifact version, and store a copy at `./sbom` on the local machine.
```

<Tabs>
<Tab heading="HCL2">

```hcl
provisioner "hcp-sbom" {
  auto_generate = true
  destination   = "./sbom"
  sbom_name     = "auto-generated-sbom"
}
```

</Tab>
<Tab heading="JSON">

```json
{
  "type": "hcp-sbom",
  "auto_generate": true,
  "destination": "./sbom",
  "sbom_name": "auto-generated-sbom"
}
```

</Tab>
</Tabs>

When `auto_generate` is enabled:
- Packer automatically detects the remote host's OS and architecture
- Downloads the appropriate Syft binary for SBOM generation (supports Linux, macOS, Windows on amd64/arm64)
- Uploads the SBOM generation tool to the remote host
- Executes the tool to generate an SBOM
- Downloads the SBOM and uploads it to HCP Packer

> - Cleans up temporary files

Reviewer: ```suggestion
When you set `auto_generated` to `true`, Packer performs the following steps, Packer automatically detects the remote host's OS and architecture and downloads the appropriate Syft binary for SBOM generation. Packer then uploads the SBOM generation tool to the remote host, executes the tool to generate an SBOM, downloads the SBOM, and uploads it to HCP Packer

After Packer completes your build, it cleans up all temporary files it created to download the SBOM generation tool.
```

<Tabs>
<Tab heading="HCL2">

```hcl
provisioner "hcp-sbom" {
  auto_generate    = true
  scanner_url      = "https://example.com/my-sbom-tool.tar.gz"
  scanner_checksum = "sha256:abc123..."
  scanner_args     = ["-o", "spdx-json", "--verbose"]
  scan_path        = "/opt/myapp"
  destination      = "./sbom"
  sbom_name        = "custom-tool-sbom"
}
```

</Tab>
<Tab heading="JSON">

```json
{
  "type": "hcp-sbom",
  "auto_generate": true,
  "scanner_url": "https://example.com/my-sbom-tool.tar.gz",
  "scanner_checksum": "sha256:abc123...",
  "scanner_args": ["-o", "spdx-json", "--verbose"],
  "scan_path": "/opt/myapp",
  "destination": "./sbom",
  "sbom_name": "custom-tool-sbom"
}
```

</Tab>
</Tabs>

### Elevated Execution Examples

#### Unix/Linux with Sudo

By default, the provisioner uses `sudo` for elevated permissions on Unix-like systems:

```hcl
provisioner "hcp-sbom" {
  auto_generate = true
  destination   = "./sbom"
  sbom_name     = "sbom-with-sudo"
  # Default execute_command includes sudo:
  # "chmod +x {{.Path}} && sudo {{.Path}} {{.Args}} {{.ScanPath}} > {{.Output}}"
}
```

> For password-based sudo:

Reviewer: ```suggestion
You can set the `execute_command` to override this default behavior. Packer provides template variables to fill in information at run time such as the path to the SBOM generation tool, arguments for the tool, and the path to scan. Refer to [Template variables in execute_command](#template-variables-in-execute_command) for a complete list of supported variables.

The following example overrides the default `execute_command` value to provide a password for the `sudo` command:
```

```hcl
provisioner "hcp-sbom" {
  auto_generate   = true
  execute_command = "chmod +x {{.Path}} && echo 'mypassword' | sudo -S {{.Path}} {{.Args}} {{.ScanPath}} > {{.Output}}"
  destination     = "./sbom"
  sbom_name       = "sbom-sudo-password"
}
```

Without sudo (if not needed):

```hcl
provisioner "hcp-sbom" {
  auto_generate   = true
  execute_command = "chmod +x {{.Path}} && {{.Path}} {{.Args}} {{.ScanPath}} > {{.Output}}"
  destination     = "./sbom"
  sbom_name       = "sbom-no-sudo"
}
```

#### Windows with Elevated User

> For Windows hosts requiring administrative privileges:

Reviewer: ```suggestion
For Windows hosts that require administrator privileges, you can set the `elevated_user` and `elevated_password` parameters to tell Packer to execute the SBOM generation as a specific user.
```
